Add a blog post: Understanding the ECMAScript spec, part 1 (#321)
v8/v8.dev: src/blog/understanding-ecma-part-1.md

To describe how `Object.prototype.hasOwnProperty` works, the spec uses pseudocode-like descriptions like this:

> `Object.prototype.hasOwnProperty(V)`
>
> When the `hasOwnProperty` method is called with argument `V`, the following steps are taken:
>
> 1. Let `P` be `? ToPropertyKey(V)`.
> 2. Let `O` be `? ToObject(this value)`.
> 3. Return `? HasOwnProperty(O, P)`.

[Spec: Object.prototype.hasOwnProperty](https://tc39.es/ecma262#sec-object.prototype.hasownproperty)

and

> `HasOwnProperty(O, P)`
>
>The abstract operation `HasOwnProperty` is used to determine whether an object has an own property with the specified property key. A Boolean value is returned. The operation is called with arguments `O` and `P` where `O` is the object and `P` is the property key. This abstract operation performs the following steps:
>
> 1. Assert: `Type(O)` is `Object`.
> 2. Assert: `IsPropertyKey(P)` is `true`.
> 3. Let `desc` be `? O.[[GetOwnProperty]](P)`.
> 4. If `desc` is `undefined`, return `false`.
> 5. Return `true`.

[Spec: HasOwnProperty](https://tc39.es/ecma262#sec-hasownproperty)

> But what's an "abstract operation"? What are the things inside \[\[ \]\]? Why is there a "?" in front of a function? What do the asserts mean?

Reviewer: ```suggestion
But what's an "abstract operation"? What are the things inside `[[ ]]`? Why is there a `?` in front of a function? What do the asserts mean?
```

Reviewer: same for other inline code spans below

[Spec: ECMAScript language types](https://tc39.es/ecma262/#sec-ecmascript-language-types)

In addition to language types, the spec also uses **specification types**, which are types that occur only in the spec, but not in the JavaScript language. The JavaScript engine does not need to (but is free to) implement them. In this blog post, we'll get to know the specification type Record (and its subtype Completion Record).

[Spec: ECMAScript specification types](https://tc39.es/ecma262/#sec-ecmascript-specification-types)

## Abstract operations

**Abstract operations** are functions defined in the ECMAScript spec; they are defined for the purpose of writing the spec concisely. A JavaScript engine doesn't have to implement them as separate functions inside the engine. They cannot be directly called from JavaScript.

[Spec: Abstract Operations](https://tc39.es/ecma262/#sec-abstract-operations)

## Internal slots and internal methods

> **Internal slots** and **internal methods** use names enclosed in \[\[ \]\].

Reviewer: ```suggestion
**Internal slots** and **internal methods** use names enclosed in `[[ ]]`.
```

> Internal slots are data members (of a JavaScript object or a specification type). They are used for storing the state of the object. Internal methods are member functions (of a JavaScript object or a specification type).

Reviewer 2: Nit: Only JS objects have internal methods. Internal methods are a pre-defined vtable that all JS objects have in the spec. Slots on the other hand are used more liberally.

Author: Ah, thanks for pointing out this error!

> 5. If `O` is an abrupt completion, return `O`
> 6. Set `O` to `O.[[Value]]`
> 7. Let `temp` be `HasOwnProperty(O, P)`.
> 8. If `temp` is an abrupt completion, return `temp`
> 9. Let `temp` be `temp.[[Value]]`
> 10. Return `NormalCompletion(temp)`

and `HasOwnProperty` like this:

> `HasOwnProperty(O, P)`
>
> 1. Assert: `Type(O)` is `Object`.
> 2. Assert: `IsPropertyKey(P)` is `true`.
> 3. Let `desc` be `O.[[GetOwnProperty]](P)`.
> 4. If `desc` is an abrupt completion, return `desc`
> 5. Set `desc` to `desc.[[Value]]`
> 6. If `desc` is `undefined`, return `NormalCompletion(false)`.
> 7. Return `NormalCompletion(true)`.

We can also rewrite the `[[GetOwnProperty]]` internal method without the exclamation mark like this:

> `O.[[GetOwnProperty]]`
>
> 1. Let `temp` be `OrdinaryGetOwnProperty(O, P)`
> 2. Assert: `temp` is not an abrupt completion
> 3. Let `temp` be `temp.[[Value]]`
> 4. Return `NormalCompletion(temp)`

Here we assume that `temp` is a brand new temporary variable which doesn't collide with anything else.

We've also used the knowledge that when a return statement returns something else than a Completion Record, it's implicitly wrapped inside a `NormalCompletion`.

> ### Side track: Return ? Foo()

Reviewer: ```suggestion
### Side track: `Return ? Foo()`
```

The spec uses the notation `Return ? Foo()` - why the question mark?

`Return ? Foo()` expands to:

> 1. Let `temp` be `Foo()`
> 2. If `temp` is an abrupt completion, return `temp`
> 3. Set `temp` to `temp.[[Value]]`
> 4. Return `NormalCompletion(temp)`

Which is the same as `Return Foo()`; it behaves the same way for both abrupt and normal completions.

## Asserts

> Asserts in the spec assert invariant conditions of the algorithms. They are added for clarity, but don't add any requirements to the implementation - the implementation doesn't need to check them.

Reviewer 2: Formatting nit: I was taught in American schools an em dash be used in situations like these.

## Moving on

We have built the understanding needed for reading the spec for simple methods like `Object.prototype.hasOwnProperty` and abstract operations like `HasOwnProperty`. They still delegate to other abstract operations, but based on this blog post we should be able to figure out what they do. We'll encounter Property Descriptors, which is just another specification type.

<figure>

>   <img src="/_img/understanding-ecma-part-1-1.svg" height="306" width="1082" alt="Function call graph starting from Object.prototype.hasOwnProperty">

Reviewer: this needs `loading=lazy` (`npm run img`)

---
title: 'Understanding the ECMAScript spec, part 1'
author: '[Marja Hölttä](https://twitter.com/marjakh), speculative specification spectator'
avatars:
  - marja-holtta
date: 2020-01-22 13:33:37
tags:
  - ECMAScript
description: 'Tutorial on reading the ECMAScript specification'
tweet: ''
---

... where we take a simple function in the spec and try to understand the notation.

## Preface

> Even if you know JavaScript, reading its language specification, the ECMAScript spec, can be pretty daunting. At least that's how I felt when I started reading it for the first time. Reading the spec from the start doesn't help either, as the spec is not designed to be read in that order, and notations are not explained before they occur.

Reviewer: > notations are not explained before they occur

They usually are, AFAICT? However, the spec is so large that you often start reading a particular section without having read the earlier chapters explaining [the notational conventions](https://tc39.es/ecma262/#sec-notational-conventions).

Author: Removed this sentence because it doesn't add much value anyway :)

Let's start with a concrete example and walk through the spec to understand it. The following code demonstrates usage of Object.prototype.hasOwnProperty:

```javascript

> let o = {'foo': 1};

Reviewer: ```suggestion
const o = {'foo': 1};
```

(per Google style guide)

o.hasOwnProperty('foo'); // true
o.hasOwnProperty('bar'); // false
```

In the example, _o_ doesn't have a property called _hasOwnProperty_, so we walk up the prototype chain and look for it. We find it in _o_'s prototype, which is _Object.prototype_.

> When trying to find out how _Object.prototype.hasOwnProperty_ is defined, I bumped into pseudocode-like descriptions like this:

Author: I unified this a bit; kept the "I" only in the first paragraph where I tell how daunted **I** was; for the rest of the blog post, I now use "we".

> Object.prototype.hasOwnProperty(V)
>
> When the hasOwnProperty method is called with argument V, the following steps are taken:
>
> 1. Let P be **?** ToPropertyKey(V).
> 2. Let O be **?** ToObject(this value).
> 3. Return **?** HasOwnProperty(O, P).

[Spec: Object.prototype.hasOwnProperty](https://tc39.es/ecma262#sec-object.prototype.hasownproperty)

and

> HasOwnProperty(O, P)
>
>The **abstract operation** HasOwnProperty is used to determine whether an object has an own property with the specified property key. A Boolean value is returned. The operation is called with arguments O and P where O is the object and P is the property key. This abstract operation performs the following steps:
>
> 1. **Assert**: Type(O) is Object.
> 2. **Assert**: IsPropertyKey(P) is true.

> > 3. Let desc be **?** O.**\[\[GetOwnProperty\]\]**(P).

Reviewer: Let's wrap these things in code spans, which both looks nicer and lets us avoid the weird escaping syntax:

```
> 1. **Assert**: `Type(O)` is `Object`.
> 2. **Assert**: `IsPropertyKey(P)` is `true`.
> 3. Let desc be `? O.[[GetOwnProperty]](P)`.
```

Author: I'm trying to maintain a style where quotes from the spec are inside >. Did you mean the code block should also be inside > (is that even possible?) or that the > should end before the code starts (confusing!)?

Author: Ah I read your comment again, I see what you mean. So not wrapping the whole thing inside a code block but individually. I might argue thought that the whole lines should be inside the code span, like. `Let desc be ...`

Author: Experimented a bit.

Option 1:

> When the hasOwnProperty method is called with argument V, the following steps are taken:
>
> `1. Let P be ? ToPropertyKey(V).`
> `2. Let O be ? ToObject(this value).`
> `3. Return ? HasOwnProperty(O, P).`

-> I think this looks ugly

Option 2:

> When the hasOwnProperty method is called with argument V, the following steps are taken:
>
> 1. Let `P` be ? `ToPropertyKey(V)`.
> 2. Let `O` be ? `ToObject(this value)`.
> 3. Return ? `HasOwnProperty(O, P)`.

-> I think this is confusing, also it's hard to decide what should be inside the code spans and what shouldn't. Let? Return? Question marks?

Option 3:

> When the hasOwnProperty method is called with argument V, the following steps are taken:
>
> 1. Let P be **?** ToPropertyKey(V).
> 2. Let O be **?** ToObject(this value).
> 3. Return **?** HasOwnProperty(O, P).

This is what I have now, the pros are that it allows using emphasis inside these blocks and it matches what the spec does. I still like this version best...

Author: Haha, I think I painted myself in a corner here :) I started changing "spec code snippets" inside the text to be inside `code spans`, and I think it's a good idea. But now they look different inside the text and inside the spec quotes. Argh.

Reviewer: I consistently use option 2, with the difference that `?` is wrapped in the code span as well. It’s not an ordinary part of an English sentence, so it should be marked accordingly. IMHO this is the only semantically correct way of representing this content in Markdown (keeping in mind what HTML it produces).

I agree that option 1 is overkill (and ugly). Option 3 is… not great, semantically speaking.

Author: I'm convinced, going with option 2.

> 4. If desc is undefined, return false.
> 5. Return true.

[Spec: HasOwnProperty](https://tc39.es/ecma262#sec-hasownproperty)

But what's an "abstract operation"? What are the things inside \[\[ \]\]? Why is there a "?" in front of a function? What do the asserts mean?

> Let's find out!

Reviewer 3: You could consider inserting a table of content here, what aspects of the spec you are going to describe here.

Reviewer: We don't support ToCs yet (see #131) but we could of course manually insert a list.

Author: Undecided about this one...

Author: I'm deciding against a table of contents, I think it would break the flow; as all those terms would be just nonsense before they're explained.

## ECMAScript Language Types and ECMAScript Specification Types

Let's start with something that looks familiar. The spec uses values such as undefined, true and false, which we already know from JavaScript. They all are **ECMAScript language values**, values of **ECMAScript language types** which the spec also defines.

The spec also uses ECMAScript language values internally, for example, an internal data type might contain a field whose possible values are true and false.

[Spec: ECMAScript language types](https://tc39.es/ecma262/#sec-ecmascript-language-types)

> **ECMAScript specification types** are types that occur only in the ECMAScript spec, but not in the JavaScript language, and which might or might not occur as internal types in a JavaScript engine. In this blog post, we'll get to know the ECMAScript specification type _Record_ (and its subtype _Completion Record_).

Reviewer 2: ```suggestion
In addition to language types and values, the specification also uses types and values occur only in the ECMAScript spec. These are called **ECMAScript specification types**, and a JavaScript engine does not need to (but is free to) implement them. In this blog post, we'll get to know the ECMAScript specification type _Record_ (and its subtype _Completion Record_).
```

Author: (Did a slightly modified version of this suggestion.)

[Spec: ECMAScript specification types](https://tc39.es/ecma262/#sec-ecmascript-specification-types)

## Abstract operations

> **Abstract operations** are functions defined in the ECMAScript spec; they are defined for the purpose of writing the spec elegantly and concisely. A JavaScript engine doesn't have to implement them as separate functions inside the engine. They cannot be directly called from JavaScript.

Author: Dropped "elegantly". Only concisely.

[Spec: Abstract Operations](https://tc39.es/ecma262/#sec-abstract-operations)

## Internal slots and internal methods

**Internal slots** and **internal methods** use names enclosed in \[\[ \]\].

> Internal slots are (assumed) data members (of a JavaScript object or a spec-internal data type) and internal methods are (assumed) member functions (of a JavaScript object or a spec-internal data type).

Author: That they're there only for spec purposes, e.g., a JavaScript object as it's represented in the JS engine might not have them. But it's confusing, so, removed (assumed).

Internal slots and methods are used in the algorithms described by the spec. Internal slots are used for storing the state of the object, and internal methods are functions associated with the object.

[Spec: Object internal methods and internal slots](https://tc39.es/ecma262/#sec-object-internal-methods-and-internal-slots)

For example, every JavaScript object has an internal slot \[\[Prototype\]\] and an internal method \[\[GetOwnProperty\]\].

> Internal slots and methods are not accessible from JavaScript, for example, you cannot access _o.\[\[Prototype\]\]_ or call _o.\[\[GetOwnProperty\]\]()_. A JavaScript engine can implement them for their own internal use, but doesn't have to.

Reviewer 2: ```suggestion
Internal slots and methods are not accessible from JavaScript. For example, you cannot access _o.\[\[Prototype\]\]_ or call _o.\[\[GetOwnProperty\]\]()_. A JavaScript engine can implement them for their own internal use, but doesn't have to.
```

> Each internal method delegates to a similarly-named abstract operation:

Reviewer 2: This is sometimes true for ordinary objects, but many internal methods are specified with algorithmic steps like abstract operations. Perhaps "Sometimes internal methods delegate to similarly-named abstract operations, such as in the case of ordinary objects' [[GetOwnProperty]]:"?

> \[\[GetOwnProperty\]\](P)
>
> When the \[\[GetOwnProperty\]\] internal method of O is called with property key P, the following steps are taken:
>
> Return ! OrdinaryGetOwnProperty(O, P).

[Spec: \[\[GetOwnProperty\]\]](https://tc39.es/ecma262/#sec-ordinary-object-internal-methods-and-internal-slots-getownproperty-p)

(We'll find out what the exclamation mark means in the next chapter.)

_OrdinaryGetOwnProperty_ is not an internal method, since it's not associated with any object; instead, the object it operates on is passed as a parameter.

> _OrdinaryGetOwnProperty_ is called "ordinary" since it operates on ordinary objects. EcmaScript objects can be either **ordinary** or **exotic**. Ordinary objects must have the default behavior for a set of methods called **essential internal methods**. If an object deviates from the default behavior, it's exotic.

Reviewer: ```suggestion
`OrdinaryGetOwnProperty` is called "ordinary" since it operates on ordinary objects. ECMAScript objects can be either _ordinary_ or _exotic_. Ordinary objects must have the default behavior for a set of methods called _essential internal methods_. If an object deviates from the default behavior, it's exotic.
```

Author: Did this suggestion *except* I kept the terms *bold* because they stand out better that way. (I'd like to be able to skim the blog post, look for a certain term and its explanation.)

The most well-known exotic object is the _Array_, since its length property behaves in a non-default way: setting the length property can remove elements from the _Array_.

Essential internal methods are the methods listed [here](https://tc39.es/ecma262/#table-5).

## Completion Records

What about the question marks and exclamation marks? To understand them, we need to look into **Completion Records**!

> A Completion Record is a "record" (or a "struct", if you're more familiar with that term) defined by the spec. A record is a data type which has a fixed set of named fields. C-like languages have structs - JavaScript doesn't, since you can always add properties into an object.

Reviewer 2: I think the "struct" comparison is more confusing than not. The spec certainly doesn't care about the memory layout of a Record's fields, for instance (or even has a notion of memory layout). I think it would be fine to simply define what ecma262 considers Records to be in this paragraph.

Again, the Completion Record is defined only for spec purposes. A JavaScript engine doesn't have to have a corresponding internal data type.

A Completion Record has three fields:

> | Name | Description |

Reviewer: Let's use `:::table-wrapper` here, similar to other tables on v8.dev.

--- | ---
| \[\[Type\]\] | One of: normal, break, continue, return, or throw. All other types except "normal" are referred to as "abrupt completions".|
| \[\[Value\]\] | the value that was produced when the completion occurred, e.g., the return value of a function or the exception (if one is thrown)|
| \[\[Target\]\] | used for directed control transfers (ignoring it for this blog post)|

[Spec: Completion Record](https://tc39.es/ecma262/#sec-completion-record-specification-type)

> Every abstract operation implicitly returns a Completion Record. Even if it looks like an abstract operation would return a simple type such as Boolean, it's implicitly wrapped into a Completion Record with the type "normal" (see [Implicit Completion Values](https://www.ecma-international.org/ecma-262/index.html#sec-implicit-completion-values)).

Reviewer 2: We're trying to change this! The implicitness doesn't work in all contexts and is generally confusing. Maybe a note that the editors are currently exploring making all completion record handling more explicit.

If an algorithm throws an exception, it means returning a Completion Record with [[Type]] throw whose [[Value]] is the exception object. We'll ignore the break, continue and return types for now.

_ReturnIfAbrupt(argument)_ means taking the following steps:

> 1. If argument is abrupt, return argument
> 2. Set argument to argument.\[\[Value\]\]

That is, we inspect a Completion Record; if it's an abrupt completion, we return immediately. Otherwise, we extract the value from the Completion Record.

[Spec: ReturnIfAbrupt](https://tc39.es/ecma262/#sec-returnifabrupt)

_ReturnIfAbrupt_ might look like a function call, but it's not. It causes the function where _ReturnIfAbrupt()_ occurs to return, not the "ReturnIfAbrupt" function itself. It behaves more like a macro in C-like languages.

_ReturnIfAbrupt_ can be used like this:

> let obj be Foo()
> // obj is a Completion Record
> ReturnIfAbrupt(obj)
> // If we're still here, obj is now the value extracted from the Completion Record
> Bar(obj)

And now the question mark comes into play: ? Foo() is equivalent to _ReturnIfAbrupt(Foo())_.

> Similarly, Let val be ! Foo() is equivalent to:

Reviewer 2: Here and elsewhere: the specese snippets like "Let val be ! Foo()" be styled differently.

> Object.prototype.hasOwnProperty(P)
>
> 1. Let P be ToPropertyKey(V).
> 2. If P is an abrupt completion, return P
> 3. Set P to P.[[Value]]
> 4. Let O be ToObject(this value).
> 5. If O is an abrupt completion, return O
> 6. Set O to O.[[Value]]
> 7. Let temp be HasOwnProperty(O, P).
> 8. If temp is an abrupt completion, return temp
> 9. Let temp be temp.\[\[Value\]\]
> 9. Return NormalCompletion(temp)

and _HasOwnProperty_ like this:

> HasOwnProperty(O, P)
>
> 1. Assert: Type(O) is Object.
> 2. Assert: IsPropertyKey(P) is true.
> 3. Let desc be O.\[\[GetOwnProperty\]\](P).
> 4. If desc is an abrupt completion, return desc
> 5. Set desc to desc.[[Value]]
> 6. If desc is undefined, return NormalCompletion(false).
> 7. Return NormalCompletion(true).

We can also rewrite the _\[\[GetOwnProperty\]\]_ internal method without the exclamation mark like this:

> O.\[\[GetOwnProperty\]\]
>
> 1. Let temp be OrdinaryGetOwnProperty(O, P)
> 2. Assert: temp is not an abrupt completion
> 3. Let temp be temp.\[\[Value\]\]
> 4. Return NormalCompletion(temp)

Here we assume that _temp_ is a brand new temporary variable which doesn't collide with anything else.

> We've also used the knowledge that when a return statement returns something else than a Completion Record, it's implicitly wrapped inside a NormalCompletion.

Reviewer 2: Sorry to harp on this point. I'd add a note here that even though the implicit rule is in place, the spec itself isn't always consistent. Often predicate-like AOs are called and their bare return values are used without unwrapping via `?` and `!`. A note to the reader that this is usually clear from context, though it is being worked on upstream.

Author: Added a note (not here but above where this "implicit completion" thing is introduced for the first time).

### Side track: Return ? Foo()

The spec uses the notation "Return ? Foo()" - why the question mark?

Return ? Foo() expands to:

> 1. Let temp be Foo()
> 2. If temp is an abrupt completion, return temp
> 3. Set temp to temp.[[Value]]
> 4. Return NormalCompletion(temp)

Which is the same as "Return Foo()"; it behaves the same way for both abrupt and normal completions.

## Asserts

Asserts in the spec assert invariant conditions of the algorithms. They are added for clarity, but don't add any requirements to the implementation - the implementation doesn't need to check them.

## Moving on

> Now we have built the understanding needed for reading the spec for simple methods and abstract operations like Object.prototype.hasOwnProperty and HasOwnProperty. They still delegate to other abstract operations, but based on this blog post you should be able to figure out what they do. You'll need to find out about Property Descriptors, which is just another ECMAScript specification type.

Reviewer: ```suggestion
Now we have built the understanding needed for reading the spec for simple methods like `Object.prototype.hasOwnProperty` and abstract operations like `HasOwnProperty`. They still delegate to other abstract operations, but based on this blog post you should be able to figure out what they do. You'll need to find out about Property Descriptors, which is just another ECMAScript specification type.
```
